components/nimbus/src/stateful/gecko_prefs.rs

```diff
@@ -335,3 +357,26 @@ pub fn query_gecko_pref_store(
         .map(|store| Value::Bool(store.pref_is_user_set(&gecko_pref)))
         .unwrap_or(Value::Bool(false)))
 }
+
+pub type MapOfExperimentSlugToPreviousState = HashMap<String, PreviousGeckoPrefState>;
+pub fn build_previous_states(states: &[GeckoPrefState]) -> MapOfExperimentSlugToPreviousState {
+    let mut original_gecko_states = MapOfExperimentSlugToPreviousState::new();
+
+    for state in states {
+        let Some(enrollment_value) = &state.enrollment_value else {
+            continue;
+        };
+
+        let original_gecko_pref = OriginalGeckoPref::from(state);
+
+        original_gecko_states
+            .entry(enrollment_value.experiment_slug.clone())
+            .and_modify(|st| st.original_values.push(original_gecko_pref.clone()))
+            .or_insert_with(|| PreviousGeckoPrefState {
+                original_values: vec![original_gecko_pref],
+                feature_id: enrollment_value.feature_id.clone(),
+                variable: enrollment_value.variable.clone(),
+            });
```

Comment:
If we end up in the the and_modify and or_insert_with cases are mutually exclusive, but unfortunately the compiler doesn't know that, otherwise we could get away without the clone. However, we can remove the clone by moving the .into() into the lambdas:

```suggestion
        original_gecko_states
            .entry(enrollment_value.experiment_slug.clone())
            .and_modify(|st| st.original_values.push(state.into()))
            .or_insert_with(|| PreviousGeckoPrefState {
                original_values: vec![state.into()],
                feature_id: enrollment_value.feature_id.clone(),
                variable: enrollment_value.variable.clone(),
            });
```

The unconditional clone() in the call to entry is unfortunate but thats the cost of using `HashMap<String, T>`.

Reply:
Thank you for the explanation!